### PR TITLE
Refinements and super

### PR DESCRIPTION
**Pull request by ssnickolay**
Closes https://github.com/oracle/truffleruby/issues/2036

We have two problems that this PR should fix.

### 1. Incomplete DeclarationContext in lookupSuperMethod

Currently, we [see two types of refinements](https://github.com/ssnickolay/truffleruby/blob/efa9b40e83f9d70f1ad677c7de9a74727e9e419f/src/main/java/org/truffleruby/core/module/ModuleOperations.java#L480):

1) Active when the method was defined

```ruby
module Test
  using M
  
  def foo
    # currentMethod.getDeclarationContext()
    # returns a context with M
    super
  end
end
```

2) The current R:

```ruby
module M
  refine C do
    def foo
      # currentMethod.getDeclarationContext()
      # returns a context with M
      super
    end
  end
end
```

It works with "flat" Rs, but if we use `include`:

```ruby
class C 
  def foo
    "should return this one"
  end
end

module A
 def foo
    # currentMethod.getDeclarationContext()
    # returns a context without any refinements
    super
  end
end

module M
  refine C do
    include A
  end
end

using M

C.new.foo
# (master TR) => NoMethodError
```

Note: we cannot solve it just forcibly looking at `objectMetaClass` (aka `C`), because we must consider other extensions (aka included modules) of the active refinements. (see test "looks in the current active refinement from included modules")

So somehow active M refinement should be available in `A#foo`.

### Solution

The first idea was to change [the last place](https://github.com/ssnickolay/truffleruby/blob/2e8c75bad33e4318f198804fde9e2158bdf37a42/src/main/java/org/truffleruby/language/dispatch/CachedDispatchNode.java#L132-L133) when we have access to the original callerFrame.

`RubyArguments.pack` extracts `DeclarationContext` from the method, but we could pass it explicitly.

The problem is in this case we will "activate" more then we should:
```ruby
module A
  def foo
    # currentMethod.getDeclarationContext()
    # returns a context without any refinements and it is correct
    bar
  end
end

class C; end

module M
  refine C do
    include A

    def bar
      "you don't see me"
    end
  end
end

using M

C.new.foo
# (MRI) => undefined local variable or method `bar'
```

In other words, `super` can see active refinements of the current class, but other (standard) methods should not.  

Thus, I changed `LookupSuperMethodNode` to pass active refinements about `selfMetaClass` (aka `C`) to the lookup method if exist.

### 1. Crazy super + refine (https://bugs.ruby-lang.org/issues/16977)

As we discussed on DM, super + refinements have not strict specifications.

```ruby
using M # where R.ancestors = R, A, B, C, Object, ...
using M1 # where R1.ancestors = R1, D, E, C, Object, ...

# If super is called from the method contained in R table methods
#    - we should ignore all other active refinements and search only in the current one:
declaringModule = R # when def foo; super end is placed
objectMetaClass = C

Serach order: [skip all before declaringModule] -> A -> B -> C -> Object -> ...

# If super is called from method included to R modules
#    - we should go through all active refinements first, and then to C:
declaringModule = D # when def foo; super end is placed
objectMetaClass = C

Serach order: [skip all before declaringModule] -> E -> R -> A -> B -> C -> Object -> ...
```

Since we are using a recursive, we need to return the variable `foundDeclaringModule` from the nested calls. To do it, I added `SuperMethodLookup` helper wrapper.

_I'm going to add more questions to the PR by place instead of listing them in the description._

### CODE REVIEW

File: src/main/java/org/truffleruby/language/FrameSendingNode.java [hunk 1]
```diff
@@ -73,6 +73,9 @@ protected synchronized void resetNeedsCallerAssumption() {
     public MaterializedFrame getFrameIfRequired(VirtualFrame frame) {
         switch (sendsFrame) {
             case MY_FRAME:
+                if (frame == null) {
```
Comment (ssnickolay) [inline src/main/java/org/truffleruby/language/FrameSendingNode.java:76]: I do not like these changes. I had to do this to prove the main idea, but let's think about how we can avoid this.

I know the reason for this:
`LookupSuperMethodNode # getDeclarationContext` executes` readCallerFrame.execute (frame) `, which is probably not pure. I do not understand, but for some reason, if we execute this code, later `getFrameIfRequired` will be executed with an empty frame.  
This happens not in the test code itself, but earlier, when mspec is initialized. If we run just empty `jt ruby test.rb`, there is not error. Maybe you have an idea why this is so happening?

File: src/main/java/org/truffleruby/language/supercall/LookupSuperMethodNode.java [hunk 2]
```diff
@@ -44,15 +50,15 @@
     protected InternalMethod lookupSuperMethodCachedDynamicObject(VirtualFrame frame, DynamicObject self,
             @Cached("getCurrentMethod(frame)") InternalMethod currentMethod,
             @Cached("metaClass(self)") DynamicObject selfMetaClass,
-            @Cached("doLookup(currentMethod, selfMetaClass)") MethodLookupResult superMethod) {
+            @Cached("doLookup(frame, currentMethod, selfMetaClass)") MethodLookupResult superMethod) {
```
Comment (ssnickolay) [inline src/main/java/org/truffleruby/language/supercall/LookupSuperMethodNode.java:53]: AFAIU we should not change anything regarding @Cached. Correct me if I'm wrong, please

File: src/main/java/org/truffleruby/language/supercall/LookupSuperMethodNode.java [hunk 3]
```diff
@@ -67,10 +73,12 @@ protected DynamicObject metaClass(Object object) {
         return metaClassNode.executeMetaClass(object);
     }
 
-    @TruffleBoundary
```
Comment (ssnickolay) [inline src/main/java/org/truffleruby/language/supercall/LookupSuperMethodNode.java:70]: I had to delete boundary annotation because we have a frame in the method arguments

File: src/main/java/org/truffleruby/core/module/ModuleOperations.java [hunk 4]
```diff
@@ -495,68 +495,78 @@ public static InternalMethod lookupMethod(DynamicObject module, String name, Vis
...
         assert RubyGuards.isRubyModule(objectMetaClass);
         final String name = currentMethod.getSharedMethodInfo().getName(); // use the original name
         return lookupSuperMethod(
                 currentMethod.getDeclaringModule(),
-                null,
+                new SuperMethodLookup(null, false, null),
                 name,
                 objectMetaClass,
-                currentMethod.getDeclarationContext());
+                declarationContext).getResult();
```
Comment (ssnickolay) [inline src/main/java/org/truffleruby/core/module/ModuleOperations.java:508]: I think an additional wrapper should not affect performance, but we can do benchmarks again (if we have thme about `super` execution)

**Comment by eregon** [comment]
The TCK tests seem to fail in CI, you can reproduce with `jt test tck`.

File: src/main/java/org/truffleruby/language/FrameSendingNode.java (hunk 1 above)
Comment (eregon) [reply src/main/java/org/truffleruby/language/FrameSendingNode.java:76]: ReadCallerFrameNode#execute() should not have visible side effects (it will just notify the caller to always send it the frame, so it doesn't have to walk the stack the next time it's used).
How can I reproduce that case?

Comment (ssnickolay) [reply src/main/java/org/truffleruby/language/FrameSendingNode.java:76]: Just comment this `if` statement, simply do

```sh
jt build
jt test fast
```

and you will see the error (not from the LookupSuperMethodNode).

Moreover, I tried change `LookupSuperMethodNode#getDeclarationContext` to
```java

private DeclarationContext getDeclarationContext(VirtualFrame frame, DynamicObject selfMetaClass) {
  final DeclarationContext context = RubyArguments.getDeclarationContext(frame);

  readCallerFrame.execute(frame); // just call it. Dont use the result

  return context;
}
```

and it has the same error;

File: spec/ruby/core/module/refine_spec.rb [hunk 5]
```diff
@@ -762,6 +787,120 @@ def foo
...
+
+      result = nil
+      Module.new do
+        using refinement
+        using refinement1
+        result = refined_class.new.foo
+      end
+
+      result.should == "foo from A"
+    end
```
Comment (eregon) [inline spec/ruby/core/module/refine_spec.rb:903]: I think these examples would be a lot easier to read if we returned an Array of the module names of the methods called, like `["b", "a", "C"]`. That would also guarantee the implementation anything in between that just does `super`.
Could you use this style in the new specs?

File: src/main/java/org/truffleruby/core/module/SuperMethodLookup.java [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    private final DynamicObject lookupTo;
+
+    public SuperMethodLookup(MethodLookupResult result, boolean foundDeclaringModule, DynamicObject lookupTo) {
+        this.result = result;
+        this.foundDeclaringModule = foundDeclaringModule;
+        this.lookupTo = lookupTo;
+    }
+
+    public SuperMethodLookup withResult(MethodLookupResult result) {
+        if (result == null) {
```
Comment (eregon) [inline src/main/java/org/truffleruby/core/module/SuperMethodLookup.java:26]: `if (result == this.result) {`

File: spec/ruby/core/module/refine_spec.rb [hunk 7]
```diff
@@ -762,6 +787,120 @@ def foo
...
+        Module.new do
+          using refinement
+          using refinement_with_super
+          refined_class.new.bar
+        end
+      }.should raise_error(NoMethodError)
+    end
+
+    it "looks in the another active refinement if super called from included modules
+        (https://bugs.ruby-lang.org/issues/16977)" do
```
Comment (eregon) [inline spec/ruby/core/module/refine_spec.rb:820]: Could you move that link as a comment above the `it`?
Multi-line or too long spec descriptions should be avoided in general.

**Comment by eregon** [comment]
Taking the simple example:
```ruby
class C
  def foo
    [:C]
  end
end

module A
  def foo
    [:A] + super
  end
end

refinement = Module.new do
  refine C do
    include A
  end
end

using refinement
p C.new.foo # => [:A, :C]
```
Real-world usage in https://github.com/ruby-next/ruby-next/blob/4945bc6ca1c976c3a3873f134168e4b6a87d71df/spec/test_unit_to_mspec.rb#L63

Thank you for the call, I think that clarified quite a few things.
As discussed, I think we should try a different approach.
The approach in this PR seems overall fine, but:
* It's expensive to get the caller frame for every call to a method with `super` in it. At least only methods with a `super` will trigger that, so that limits it quite a bit. Could potentially pass a separate DeclarationContext for `super` lookup in CachedDispatchNode#call, but:
* My understanding (might be wrong) is refinements are purely lexical, so we shouldn't need to pass which refinements are active when calling `.foo` (so that `foo` can know which super method to call), but rather on `include A` inside `refine C` we could copy the methods of A (and A's ancestors) and give them `C => R` as active refinements (as part of the InternalMethod DeclarationContext), as if `def foo` was inside `refine C` without the `include`.

As we both found out modules included in refinements are not well defined together with `super`:
* https://bugs.ruby-lang.org/issues/16977 (mentioned above, shows inconsistency when using `include` vs inlined)
* https://bugs.ruby-lang.org/issues/17007 (new, SystemStackError)

So I'd like to focus on getting reasonable cases first, like the one mentioned above, and not necessarily reproduce potential bugs in CRuby.

File: src/main/java/org/truffleruby/language/supercall/LookupSuperMethodNode.java [hunk 8]
```diff
@@ -44,15 +49,15 @@
     protected InternalMethod lookupSuperMethodCachedDynamicObject(VirtualFrame frame, DynamicObject self,
```
Comment (eregon) [inline src/main/java/org/truffleruby/language/supercall/LookupSuperMethodNode.java:49]: Note: the guards would need to check that the caller frame/parent's DeclarationContext is the same, as it could change if the Module is included in multiple refinements (I wonder if that actually works on MRI).

File: src/main/java/org/truffleruby/language/FrameSendingNode.java (hunk 1 above)
Comment (ssnickolay) [reply src/main/java/org/truffleruby/language/FrameSendingNode.java]: Not relevant more

File: src/main/java/org/truffleruby/core/module/ModuleOperations.java [hunk 9]
```diff
@@ -569,6 +571,10 @@ private static SuperMethodLookup lookupSuperMethod(DynamicObject declaringModule
         return lookup.withResult(new MethodLookupResult(null, toArray(assumptions)));
     }
 
+    private static InternalMethod rememberUsedRefinement(InternalMethod method, DeclarationContext declarationContext) {
```
Comment (eregon) [inline src/main/java/org/truffleruby/core/module/ModuleOperations.java:574]: Maybe rememberUsedRefinement*s* since there might be multiple active refinements?

File: spec/ruby/core/module/refine_spec.rb [hunk 10]
```diff
@@ -760,7 +785,140 @@ def foo
...
+        end
+      end
+
+      -> {
+        Module.new do
+          using refinement
+          using refinement_with_super
+          refined_class.new.bar
+        end
+      }.should raise_error(NoMethodError)
```
Comment (eregon) [inline spec/ruby/core/module/refine_spec.rb:816]: `should raise_error` should be used on the narrowest scope possible, so here around `instance.bar`

File: src/main/java/org/truffleruby/core/module/ModuleOperations.java [hunk 11]
```diff
@@ -495,68 +496,84 @@ public static InternalMethod lookupMethod(DynamicObject module, String name, Vis
...
                 name,
                 objectMetaClass,
-                currentMethod.getDeclarationContext());
+                declarationContext).getResult();
     }
 
+
     @TruffleBoundary
-    private static MethodLookupResult lookupSuperMethod(DynamicObject declaringModule, DynamicObject lookupTo,
+    private static SuperMethodLookup lookupSuperMethod(DynamicObject declaringModule, SuperMethodLookup lookup,
```
Comment (eregon) [inline src/main/java/org/truffleruby/core/module/ModuleOperations.java:514]: Do we need an extra object SuperMethodLookup here, or could we just pass more arguments?
The call above only extracts the results, and the recursive calls could pass extra arguments.
We could have a Memo<Boolean> or boolean[1] for `foundDeclaringModule` if it needs to be modified by recursive calls and read in the caller.

I see this method already has many arguments, but if that's the worry maybe we should encapsulate all of them in some arguments object. I tend to prefer explicit arguments because it's clearer what are the inputs.

File: src/main/java/org/truffleruby/language/methods/InternalMethod.java [hunk 12]
```diff
@@ -173,6 +178,10 @@ public boolean isBuiltIn() {
         return builtIn;
     }
 
+    public boolean isRefined() {
+        return activeRefinements != null;
+    }
```
Comment (eregon) [inline src/main/java/org/truffleruby/language/methods/InternalMethod.java:183]: This sounds a bit confusing as it doesn't mean all refined method will return `true` here, only if they come from lookupMethod(). I would do `getActiveRefinements() != null` in callers instead.

File: src/main/java/org/truffleruby/language/supercall/LookupSuperMethodNode.java (hunk 3 above)
Comment (eregon) [reply src/main/java/org/truffleruby/language/supercall/LookupSuperMethodNode.java:70]: We should pass the `DeclarationContext` context and not the entire frame, and keep the `@TruffleBoundary` here.

File: src/main/java/org/truffleruby/language/supercall/LookupSuperMethodNode.java [hunk 13]
```diff
@@ -82,4 +90,27 @@ protected int getCacheLimit() {
         return getContext().getOptions().METHOD_LOOKUP_CACHE;
     }
 
+    private DeclarationContext getDeclarationContext(VirtualFrame frame, InternalMethod currentMethod,
```
Comment (eregon) [inline src/main/java/org/truffleruby/language/supercall/LookupSuperMethodNode.java:93]: Should also not receive the frame because it includes HashMap operations below.
Callers can pass `RubyArguments.getDeclarationContext(frame)` instead of the frame itself.
And then this method can be @TruffleBoundary as well.

File: src/main/java/org/truffleruby/language/supercall/LookupSuperMethodNode.java [hunk 14]
```diff
@@ -82,4 +90,27 @@ protected int getCacheLimit() {
...
+            final DynamicObject[] classRefinements = currentMethod
+                    .getActiveRefinements()
+                    .getRefinementsFor(selfMetaClass);
+
+            if (classRefinements == null) {
+                return context;
+            } else {
+                // add to the context active refinements for the selfMetaClass
+                final Map<DynamicObject, DynamicObject[]> newRefinements = new HashMap<>(context.getRefinements());
+                newRefinements.put(selfMetaClass, classRefinements);
```
Comment (eregon) [inline src/main/java/org/truffleruby/language/supercall/LookupSuperMethodNode.java:108]: That seems quite expensive. Do we need to look into the `RubyArguments.getDeclarationContext(frame)` or just `classRefinements` is enough to find the super method?

If we need it, I don't see much other way, so probably fine.
Would be worth checking if `context.getRefinements()` is empty then, in which case we don't need a new Map.

File: spec/ruby/core/module/refine_spec.rb [hunk 15]
```diff
@@ -760,7 +785,142 @@ def foo
...
+          end
+        end
+      end
+
+
+      Module.new do
+        using refinement
+        -> {
+          refined_class.new.foo
+        }.should raise_error(NameError)
```
Comment (eregon) [inline spec/ruby/core/module/refine_spec.rb:844]: It would be good to verify it's indeed `bar` missing here (so the spec doesn't pass if e.g. `foo` would raise a NameError due to some bug), with:
```ruby
        }.should raise_error(NameError) { |e| e.name.should == :bar }
```
Same for the spec above.

File: spec/ruby/core/module/refine_spec.rb [hunk 16]
```diff
@@ -760,7 +785,142 @@ def foo
...
+      refinement1 = Module.new do
+        refine refined_class do
+          include b
+        end
+      end
+
+      result = nil
+      Module.new do
+        using refinement
+        using refinement1
```
Comment (eregon) [inline spec/ruby/core/module/refine_spec.rb:879]: refinement1 & refinement2 would be clearer names here (or refinement_a & refinement_b)

File: doc/contributor/refinements.md [hunk 17]
```diff
@@ -88,12 +88,15 @@ The lookup order:
...
 
-The lookup order for `super`:
-```ruby
-R1 -> A -> B -> C -> ... -> R2 -> D -> E -> C -> ...
-```
+
+The `super` lookup [works in two modes](https://bugs.ruby-lang.org/issues/16977):
+
+1. If `super` is called from a method is directly in R, then we should search in `C` ancestors and ignore other active refinements.
+2. If `super` is called from a method placed in a module which included to R, then we should search overall active refinements (as we do for a regular lookup).
```
Comment (eregon) [inline doc/contributor/refinements.md:97]: `overall` => `all` (or `over all`)

**Comment by eregon** [comment]
FYI I ran benchmarks on cb1a27c53c0 "tmp: keep parent declaration context in RubyArguments" (https://github.com/oracle/truffleruby/compare/master...cb1a27c53c02c9011f0330eb4d19ec712b45d3f1) and I can observe a slowdown on `binary-trees` and `red-black` of 8-9%, they don't use `super` but they do lots of recursive so the extra argument probably causes that slowdown.

File: src/main/java/org/truffleruby/core/module/ModuleOperations.java (hunk 11 above)
Comment (ssnickolay) [reply src/main/java/org/truffleruby/core/module/ModuleOperations.java:514]: Is the complexity is the reason why you want to change it? It probably should not increase footprint, but maybe affect performace a little bit since we allocate a couple of additional `SuperMethodLookup` objects. 

The reason why I added extra wrapper `SuperMethodLookup` is I don't like\prefer methods that return the result, but at the same time modifying something "hidden" (in our case return `MethodLookupResult` and modify `foundDeclaringModule`). It’s good when one of the two is used (it is advisable to explicitly work with the returned result), but not both at once.

However, `Memo<boolean>` It looks fine to me because it "screams" about its purpose to keep something between calls (I just did not know about this class).

**Comment by ssnickolay** [comment]
> FYI I ran benchmarks on the old approach cb1a27c "tmp: keep parent declaration context in RubyArguments" (master...cb1a27c) and I can observe a slowdown on binary-trees and red-black of 8-9%, they don't use super but they do lots of recursive calls so the extra argument probably causes that slowdown.

Thank you for sharing with me! I'm surprised that the storage of just one additional argument has such a big impact. Glad we found a better solution

File: src/main/java/org/truffleruby/core/module/ModuleOperations.java (hunk 11 above)
Comment (eregon) [reply src/main/java/org/truffleruby/core/module/ModuleOperations.java:514]: My main concern here is readability, I find lookupSuperMethod() very hard to follow with both MethodLookupResult and SuperMethodLookup. One possibility would be to have SuperMethodLookup replaces the MethodLookupResult for lookupSuperMethod() methods, so we'd have only one.
But I think the Memo caries the intention more clearly: it's state only useful for recursion, not for the caller.

File: src/main/java/org/truffleruby/language/supercall/LookupSuperMethodNode.java (hunk 14 above)
Comment (ssnickolay) [reply src/main/java/org/truffleruby/language/supercall/LookupSuperMethodNode.java:110]: > Do we need to look into the RubyArguments.getDeclarationContext(frame)

Yeah, we need it because it contains the method refinements:

```ruby
module A
  using SomeM
  
  def foo
    super + ??? # here we should see SomeM; aka RubyArguments.getDeclarationContext(frame)
  end
end

refine C
  include A # pass classRefinements
end
```

However, we know that `RubyArguments.getDeclarationContext(frame) ==  currentMethod.getDeclarationContext()` (it what RubyArguments.pack does) so we can replace one with another.

Additionally, it helps to remove `frame` arguments and return `@TruffleBoundary` 

See: https://github.com/oracle/truffleruby/pull/2039/commits/d1cdef8ff843f3e87d0a07449b7ac9a3d8e02561

File: src/main/java/org/truffleruby/language/methods/InternalMethod.java (hunk 12 above)
Comment (ssnickolay) [reply src/main/java/org/truffleruby/language/methods/InternalMethod.java]: Part of https://github.com/oracle/truffleruby/pull/2039/commits/6b515da1daaa15c12b505ac0916f404f76b3fcf3

File: src/main/java/org/truffleruby/language/supercall/LookupSuperMethodNode.java (hunk 14 above)
Comment (eregon) [reply src/main/java/org/truffleruby/language/supercall/LookupSuperMethodNode.java:109]: Do we have a spec that tests this, that `super` will also use refinements in its lexical scope (not just refinements when looking up the method which has `super` in it)?
Semantically it might make more sense to consider only one set of refinements for `super`, so from the first call, it's a well-defined way up the chain. But of course we should match MRI.

Comment (ssnickolay) [reply src/main/java/org/truffleruby/language/supercall/LookupSuperMethodNode.java]: Done: https://github.com/oracle/truffleruby/pull/2039/commits/05e27ecd9dda8891d184c553fdaea7bb03e5c61c
